support aws ebs snapshot based backup and restore (#2065)
pingcap/docs-tidb-operator: en/restore-from-aws-s3-by-snapshot.md

---
title: Restore a TiDB Cluster from EBS Volume Snapshots
summary: Learn how to restore backup metadata and EBS volume snapshots from S3 storage to a TiDB cluster.
---

# Restore a TiDB Cluster from EBS Volume Snapshots

This document describes how to restore backup data in AWS EBS snapshots from S3 storage to a TiDB cluster.

The restore method described in this document is based on TiDB Operator's CustomResourceDefinition (CRD). The backup data based on AWS EBS snapshots contains two parts, the snapshots of the TiDB cluster data volumes, and the backup metadata of the snapshots and the cluster.

## Requirements

> - Snapshot restore is only applicable to TiDB Operator v1.4.0 and later, and TiDB v6.3.0 and later.

Reviewer: ```suggestion
- Snapshot restore is applicable to TiDB Operator v1.4.0 or above, and TiDB v6.3.0 or above.
```

> - Snapshot restore only supports restoring to a cluster with the same number of TiKV nodes and volumes. That is, the number of TiKV nodes and volume configurations in the restore cluster must be the same as those in the backup cluster.

Reviewer: ```suggestion
- Snapshot restore only supports restoring to a cluster with the same number of TiKV nodes and volumes configuration. That is, the number of TiKV nodes and volume configurations is identical between the restore cluster and backup cluster.
```

> - Snapshot restore is currently not supported for TiFlash, TiCDC, DM, and TiDB Binlog.

Reviewer 2: ```suggestion
- Snapshot restore is currently not supported for TiFlash, TiCDC, DM, and TiDB Binlog nodes.
```

## Step 1. Set up the environment for EBS volume snapshot restore

Before using TiDB Operator to restore backup metadata and EBS snapshots from S3 storage to TiDB, prepare the restore environment by following the steps below:

1. Download the file [backup-rbac.yaml](https://github.com/pingcap/tidb-operator/blob/master/manifests/backup/backup-rbac.yaml).

2. Create the RBAC-related resources required for the restore in the `test2` namespace by running the following command:

    ```shell
    kubectl apply -f backup-rbac.yaml -n test2
    ```

3. Grant permissions to access remote storage.

>     To restore data from Amazon S3, you need to grant the permissions to remote storage. Three ways are available. See [AWS account permissions](grant-permissions-to-remote-storage.md#aws-account-permissions).

Reviewer: ```suggestion
    To restore data from Amazon S3, you need to grant permission to remote storage. Three ways are available. See [AWS account permissions](grant-permissions-to-remote-storage.md#aws-account-permissions).
```

## Step 2. Prepare the restore cluster

Deploy a cluster to which you want to restore data. See [Deploy TiDB on AWS EKS](deploy-on-aws-eks.md).

> Add the `recoveryMode: true` field to Spec and run the following command to create the resources required for the restore in the `test2` namespace:

Reviewer 2: ```suggestion
Add the `recoveryMode: true` field to spec and run the following command to create the resources required for the restore in the `test2` namespace:
```

```shell
kubectl apply -f tidb-cluster.yaml -n test2
```

## Step 3. Restore backup data to the TiDB cluster

> Based on the authorization method you selected in the previous step to grant remote storage access, you can restore data to TiDB using any of the following methods accordingly:

Reviewer 2: ```suggestion
Based on the authorization method you selected in [step 1](#step-1-set-up-the-environment-for-ebs-volume-snapshot-restore) to grant remote storage access, you can restore data to TiDB using any of the following methods accordingly:
```

> + Method 1: If you authorize permission by accessKey and secretKey, you can create the `Restore` CR as follows:

Reviewer 2: ```suggestion
+ Method 1: If you grant permissions by accessKey and secretKey, you can create the `Restore` CR as follows:
```

    ```shell
    kubectl apply -f restore-aws-s3.yaml
    ```

    The `restore-aws-s3.yaml` file has the following content:

    ```yaml
    ---
    apiVersion: pingcap.com/v1alpha1
    kind: Restore
    metadata:
      name: demo2-restore-s3
      namespace: test2
    spec:
      br:
        cluster: demo2
        clusterNamespace: test2
        backupType: full
        restoreMode: volume-snapshot
        # logLevel: info
      s3:
        provider: aws
        secretName: s3-secret
        region: us-west-1
        bucket: my-bucket
        prefix: my-folder
    ```

> + Method 2: If you authorize permission by associating Pod with IAM, you can create the `Restore` CR as follows:

Reviewer 2: ```suggestion
+ Method 2: If you grant permissions by associating Pod with IAM, you can create the `Restore` CR as follows:
```

    ```shell
    kubectl apply -f restore-aws-s3.yaml
    ```

    The `restore-aws-s3.yaml` file has the following content:

    ```yaml
    ---
    apiVersion: pingcap.com/v1alpha1
    kind: Restore
    metadata:
      name: demo2-restore-s3
      namespace: test2
      annotations:
        iam.amazonaws.com/role: arn:aws:iam::123456789012:role/user
    spec:
      br:
        cluster: demo2
        sendCredToTikv: false
        clusterNamespace: test2
        backupType: full
        restoreMode: volume-snapshot
        # logLevel: info
      s3:
        provider: aws
        region: us-west-1
        bucket: my-bucket
        prefix: my-folder
    ```

> + Method 3: If you authorize permission by associating ServiceAccount with IAM, you can create the `Restore` CR as follows:

Reviewer 2: ```suggestion
+ Method 3: If you grant permissions by associating ServiceAccount with IAM, you can create the `Restore` CR as follows:
```
